### PR TITLE
TensorFlow training/inference optimization

### LINKED ISSUE #6771: Exported TF Bert model is much slower than that exported from Google's Bert
I benchmarked the Bert model exported from Huggingface TF Bert code and observed that it is much slower than Google's TF2 Bert model(https://github.com/tensorflow/models/tree/master/official/nlp/bert) on TF Serving (GPU device). When batch_size=1, Huggingface Bert model can be more than 200% slower than Google's one. Here are my benchmark results:

**latency: (ms/batch) on 1 GPU-V100**
Batch Size | google official | HF_TF2 | relative change |
-- | -- | -- | -- 
1 | 6.7 | 21.3 | 238.10%
2 | 9.4 | 24.2 | 171.91% 
4 | 14.4 | 28.1 | 99.29% 
8 | 24.0 | 36.9 | 55.70% 
16 | 46.6 | 58.6 | 26.57%
64 | 171.5 | 171.4 | 0.35% 
128 | 338.5 | 324.5 | -3.71% 

I used the latest Tensorflow profiling tool (https://www.tensorflow.org/tfx/serving/tensorboard) to compare these two:
 
**Hugging-face Bert-Base FP32 Batch_size = 128**
Screen Shot 2020-06-05 at 8 20 34 PM
**Google Bert-Base FP32 Batch_size = 128**
image (1)

Comparing the event traces of HF's and google's, we see that before each *MatMul* operation on GPU, there are always some CPU processes running (wasn’t able to get more info about these CPU processes because the Profiling tool doesn’t provide meaningful labels or description for those CPU processes but it seems like they are doing some gather/reduce operations) and the GPU is inactive for a while. If you look at the trace of google official, you can find that all GPU ops are compact and there are no idle time for GPU. In parallel, only one CPU process keeps running and doesn’t cause GPU ops to wait. The short idle time will hurt the performance more when batch_size or model_size is small because short time is spent on GPU calculations and idle time is not ignorable anymore. See time comparison grouped by type below:

**Hugging-face Bert-Base FP32 Batch_size = 4**  
image (3)

**Google Bert-Base FP32 Batch_size = 4**
image (4)

Under small batch_size=4, Hugging-face model has 60% GPU idle time while that of google model is only 25%. This also explains the reason why for small batch_size, there is over 200% slow down and for large batch_size=128, the situation gets better. 

The *MatMul* op is triggered by tf.keras.layers.Dense(), which is widely used in Transformer encoder self-attention, intermediate layer and output layer. In comparison, Google's Bert uses DenseEinsum()([link](https://github.com/tensorflow/models/blob/1ebff962832ec8a599bd67ac8c43b9ae1b05c3fa/official/nlp/transformer/attention_layer.py#L59-L69)) to replace all usage of Dense layer. I personally modified Huggingface's TF Bert code base to use DenseEinsum(). And the slow down issue got solved:

**latency: (ms/batch)  on 1 GPU-V100**
Batch Size | google official | HF_TF2 | relative change | After Fixing | relative change
-- | -- | -- | -- | -- | --
1 | 6.7 | 21.3 | 238.10% | 6.60 | 4.76%
2 | 9.4 | 24.2 | 171.91% | 8.90 | 0.00%
4 | 14.4 | 28.1 | 99.29% | 13.40 | -4.96%
8 | 24.0 | 36.9 | 55.70% | 22.10 | -6.75%
16 | 46.6 | 58.6 | 26.57% | 43.20 | -6.70%
64 | 171.5 | 171.4 | 0.35% | 158.80 | -7.03%
128 | 338.5 | 324.5 | -3.71% | 313.10 | -7.09%

Profile after fixing:
image (4) copy

I noticed that multiple issues may be related to this (https://github.com/huggingface/transformers/issues/6264).

Do you plan to solve this issue? Thanks. @patrickvonplaten

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR fixes a performance issue where some operation was done on CPU instead of GPU and would result to put the GPU in idle mode. This optimization is feasible thanks to the recent update we made on the way we load the TF weights.

@patrickvonplaten I have done few changes in the `TFLongformer` model but I'm sure it can be further optimized the same way (see `TFLongformerSelfAttention`) but as I don't know much on how works this model, can you take a look if the same optimization can be applied?

Fixes #6771

### CODE REVIEW

**Review by LysandreJik — COMMENTED** [review]
If that's all it takes, that's fantastic! Did you manage to obtain the performance improvements that were initially mentioned thanks to this?

Also I'm realizing now that we don't have integration testing for our TensorFlow models, and this seems like a situation where having some would be needed. Could we work on adding these tests for the models modified here at first, and then add it to the rest of the models?

Something like is done [in `tests/test_modeling_roberta.py`](https://github.com/huggingface/transformers/blob/master/tests/test_modeling_roberta.py#L397-L448), using tiny models.

I can help you work on it if you're lacking time!

**Comment by jplu** [comment]
> If that's all it takes, that's fantastic! Did you manage to obtain the performance improvements that were initially mentioned thanks to this?

On my machine with my GPU yes.

>Also I'm realizing now that we don't have integration testing for our TensorFlow models, and this seems like a situation where having some would be needed. Could we work on adding these tests for the models modified here at first, and then add it to the rest of the models?

Sure! It is a good idea!

> I can help you work on it if you're lacking time!

I would appreciate if you have time yes 😃

**Comment by LysandreJik** [comment]
Okay, will take a look at doing the integrations tests sometimes tonight. Will let you know!

**Comment by chiapas** [comment]
@jplu 

For learning purpose, I am wondering which operations was done on CPU instead of GPU. I saw you changed `Dense` to `EinsumDense` in several places, and remove several operations about shape changing. Is shape changing done on CPU and `EinsumDense` could avoid this? Could you give me some information about this, so I can read and learn it? Thanks.

**Comment by jplu** [comment]
@chiapas 

If you take a look at #6771 is it quite well detailed. The issue was coming from transpose+matmul that was done on CPU. einsumDense allows you to do all these computation directly in the layer but at the cost of changing the shapes of the original layers, that why we have modified the way we load the TF models.

To do this PR I basically took example on the original BERT implementation right [here](https://github.com/tensorflow/models/blob/master/official/nlp/transformer/attention_layer.py).

**Comment by jplu** [comment]
Thanks a lot @LysandreJik !!

As I'm currently working on from scratch LM training for TF models, I don't have much time to really focus on this.

**Comment by chiapas** [comment]
> transpose+matmul

@jplu  Thanks. I am superised by this `transpose+matmul that was done on CPU`.

**Comment by chiapas** [comment]
> 
> 
> Thanks a lot @LysandreJik !!
> 
> As I'm currently working on from scratch LM training for TF models, I don't have much time to really focus on this.

@jplu You also works on LM training for TF models? I plan to go back to a pending PR #6955 I created once the `test_trainer_tf.py` is done. Do PR #6955 and your work on TF models LM training overlap?

**Comment by jplu** [comment]
@chiapas This is exactly what I'm doing, and the models needs some rework that's why I'm mostly focus on BERT to have at least one model working.

**Comment by chiapas** [comment]
> @chiapas This is exactly what I'm doing, and the models needs some rework that's why I'm mostly focus on BERT to have at least one model working.
> 
> I just done yesterday the data pipeline with random masking generation.

Ah, ok. I guess my PR was pending too long and my bad not communicating with you first. I planed to do this while I finished a notebook on Kaggle [Masked, My Dear Watson - MLM with TPU](https://www.kaggle.com/yihdarshieh/masked-my-dear-watson-mlm-with-tpu), which also works on MLM.

Since you already have more progresses (and also you are HF member), it is better for you to continue.  However, if there is something I can contribute for this TF LM task, I would love to do it.

**Comment by jplu** [comment]
> Since you already have more progresses (and also you are HF member), it is better for you to continue. However, if there is something I can contribute for this TF LM task, I would love to do it.

Thanks! I will let you know.

**Comment by patrickvonplaten** [comment]
That's awesome! I will see what results the TF benchmark scripts give before/after this PR.

**Comment by patrickvonplaten** [comment]
I ran the benchmarks: `python examples/benchmarking/run_benchmark_tf.py --models bert-base-cased --env_print`  in the following environment:

```
- transformers_version: 3.3.1                                                                                                                                                                                      
- framework: TensorFlow                                                                                                                                                                                            
- eager_mode: False                                                                                                                                                                                                
- use_xla: False                                                                                                                                                                                                   
- framework_version: 2.3.0                                                                                                                                                                                         
- python_version: 3.6.10                                                                                                                                                                                           
- system: Linux                                                                                                                                                                                                    
- cpu: x86_64                                                                                                                                                                                                      
- architecture: 64bit                                                                                                                                                                                              
- date: 2020-10-06                   
- time: 19:06:48.378935                                                                                  
- fp16: False                                                                                                                                                                                                      
- use_multiprocessing: True                                                                                                                                                                                        
- only_pretrain_model: False                                                                                                                                                                                       
- cpu_ram_mb: 32088                                 
- use_gpu: True                                                                                          
- num_gpus: 1                                                                                            
- gpu: TITAN RTX                                                                                         
- gpu_ram_mb: 24217                                                                                      
- gpu_power_watts: 280.0                                                                                 
- gpu_performance_state: 8                                                                               
- use_tpu: False 
```


Currently, on master:

```
====================       INFERENCE - SPEED - RESULT       ====================                                                                                                                                   
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
       bert-base-cased               8               8             0.085     
       bert-base-cased               8               32            0.166     
       bert-base-cased               8              128            0.513     
       bert-base-cased               8              512            2.629     
--------------------------------------------------------------------------------                                                                                                                                 
```

In this `tf-optim` branch, the results are:

```
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
       bert-base-cased               8               8             0.088     
       bert-base-cased               8               32            0.176     
       bert-base-cased               8              128            0.531     
       bert-base-cased               8              512            3.028     
--------------------------------------------------------------------------------
```

=> So the speed results are more or less identical with the way the benchmarks are used.

I don't compile the model with Keras, but just add the "@tf.function" decorator to the function to transform the function into graph mode. So not sure what to think of that.... => @jplu - colud you maybe check the benchmark script and see if you can get a speed-up there? Or if the benchmark script is wrong? 

```
python examples/benchmarking/run_benchmark_tf.py --models bert-base-cased --env_print
```

**Comment by jplu** [comment]
The benchmark script is ok, but to see the difference you have to create a saved_model and run the model in TF Serving. Your benchmark don't take into account all the optimization TF serving does for inference.

We should update the benchmark script to include:

- Saved model creation
- run a the saved model with the TF Serving tool
- adapt the the benchmark to include gRPC calls to use the model from TF Serving.

**Comment by jplu** [comment]
Will be integrated into the PR #7753
